File: shared/ui/ipc/webview.protocol.extension.ts
import { NotificationType } from "vscode-jsonrpc";
import { Range } from "vscode-languageserver-types";

type ConfigState = any;

/* The following events are expected to be provided from the extension */

export interface DidHighlightCodeNotification {
	code: string;
	file: string | undefined;
	fileUri: string;
	range: Range;
	source:
		| {
				file: string;
				repoPath: string;
				revision: string;
				authors: {
					id: string;
					username: string;
				}[];
				remotes: {
					name: string;
					url: string;
				}[];
		  }
		| undefined;
	gitError: string | undefined;
	isHighlight?: boolean;
}
export const DidHighlightCodeNotificationType = new NotificationType<
	DidHighlightCodeNotification,
	void
>("webview/code-highlighted");

export interface DidSelectStreamThreadNotification {
	streamId: string;
	threadId?: string;
}

export const DidSelectStreamThreadNotificationType = new NotificationType<
	DidSelectStreamThreadNotification,
	void
>("webview/stream-thread-selected");

export interface DidScrollEditorNotification {
	uri: any; // vscode Uri type
	firstLine: number;
	lastLine: number;
}

export const DidScrollEditorNotificationType = new NotificationType<
	DidScrollEditorNotification,
	void
>("webview/text-editor-scrolled");

export const DidChangeDataNotification = new NotificationType<{ type: string; data: any }, void>(
	"webview/data-changed"
);

export type DidChangeConfigsNotification = Partial<ConfigState>;

export const DidChangeConfigsNotificationType = new NotificationType<
	DidChangeConfigsNotification,
	void
>("webview/configs-changed");

export const DidLoseConnectivityNotificationType = new NotificationType<void, void>(
	"webview/connectivity-lost"
);

export const DidEstablishConnectivityNotificationType = new NotificationType<void, void>(
	"webview/connectivity-established"
);

export interface DidChangeActiveEditorNotification {
	editor?: { fileName: string; fileStreamId?: string };
}

export const DidChangeActiveEditorNotificationType = new NotificationType<
	DidChangeActiveEditorNotification,
	void
>("webview/active-editor-changed");

export const DidFocusNotificationType = new NotificationType("webview/focused");
export const DidBlurNotificationType = new NotificationType("webview/blurred");

export const DidSignOutNotificationType = new NotificationType("webview/signed-out");

File: vs/src/CodeStream.VisualStudio/Models/Webview.Protocol.Extension.cs
using System.Collections.Generic;
using CodeStream.VisualStudio.Extensions;
using Microsoft.VisualStudio.LanguageServer.Protocol;
using Newtonsoft.Json.Linq;

namespace CodeStream.VisualStudio.Models
{
    public class DidScrollEditorNotificationTypeParams
    {
       public string Uri { get; set; }
       public List<Range> VisibleRanges { get; set; }
    }

    public class DidScrollEditorNotificationType : NotificationType<DidScrollEditorNotificationTypeParams>
    {
        public const string MethodName = "webview/text-editor-scrolled";
        public override string Method => MethodName;
    }

    public class DidChangeDataNotificationTypeParams { }
    public class DidChangeDataNotificationType : NotificationType<DidChangeDataNotificationTypeParams>
    {
        public const string MethodName = "webview/data-changed";

        private readonly JToken _token;

        public DidChangeDataNotificationType(JToken token)
        {
            _token = token;
        }

        public override string Method => MethodName;

        public override string AsJson()
        {
            return @"{""method"":""" + Method + @""",""params"":" + _token.ToJson() + "}";
        }
    }

    public class DidChangeConfigsNotificationTypeParams
    {
        public bool? MuteAll { get; set; }
        public bool? OpenCommentOnSelect { get; set; }
        public bool? ShowMarkers { get; set; }
        public bool? ViewCodemarksInline { get; set; }
    }

    public class DidChangeConfigsNotificationType : NotificationType<DidChangeConfigsNotificationTypeParams>
    {
        public const string MethodName = "webview/configs-changed";
        public override string Method => MethodName;
    }

    public class DidLoseConnectivityNotificationType : NotificationType<EmptyRequestTypeParams>
    {
        public static string MethodName = "webview/connectivity-lost";
        public override string Method => MethodName;
    }

    public class DidEstablishConnectivityNotificationType : NotificationType<EmptyRequestTypeParams>
    {
        public static string MethodName = "webview/connectivity-established";
        public override string Method => MethodName;
    }

    public class DidChangeActiveEditorNotificationTypeParamsEditor
    {
        public string FileStreamId { get; set; }
        public string Uri { get; set; }
        public string FileName { get; set; }
        public List<Range> VisibleRanges { get; set; }
    }

    public class DidChangeActiveEditorNotificationParams
    {
        public DidChangeActiveEditorNotificationTypeParamsEditor Editor { get; set; }
    }

    public class DidChangeActiveEditorNotificationType : NotificationType<DidChangeActiveEditorNotificationParams>
    {
        public static string MethodName = "webview/active-editor-changed";
        public override string Method => MethodName;
    }

    public class DidFocusNotificationType : NotificationType<EmptyRequestTypeParams>
    {
        public static string MethodName = "webview/focused";
        public override string Method => MethodName;
    }
    public class DidBlurNotificationType : NotificationType<DidChangeActiveEditorNotificationParams>
    {
        public static string MethodName = "webview/blurred";
        public override string Method => MethodName;
    }
    public class DidSignOutNotificationType : NotificationType<DidChangeActiveEditorNotificationParams>
    {
        public static string MethodName = "webview/signed-out";
        public override string Method => MethodName;
    }
}
